aiida_lsmo/workchains/cp2k_phonopy.py

```diff
@@ -0,0 +1,188 @@
+
+from aiida.orm import QueryBuilder, Node
+from aiida.plugins import CalculationFactory, DataFactory, WorkflowFactory
+from aiida.common import AttributeDict
+from aiida.engine import append_, while_, WorkChain, ToContext
+from aiida_lsmo.utils import aiida_dict_merge
+
+# import sub-workchains
+Cp2kBaseWorkChain = WorkflowFactory('cp2k.base')  # pylint: disable=invalid-name
+
+# import aiida data
+Dict = DataFactory('dict')  # pylint: disable=invalid-name
+CifData = DataFactory('cif')  # pylint: disable=invalid-name
+StructureData = DataFactory('structure')  # pylint: disable=invalid-name
+
+
+class Cp2kPhonopyWorkChain(WorkChain):
+    """A workchain to compute frequencies using CP2K and Phonopy"""
+
+    @classmethod
+    def define(cls, spec):
+        """Define workflow specification."""
+        super().define(spec)
+
+        spec.input('structure', valid_type=(CifData, StructureData), required=True, help='Input structure')
+        spec.expose_inputs(Cp2kBaseWorkChain,
+                           namespace='cp2k_base',
+                           exclude=['cp2k.structure', 'cp2k.parameters', 'cp2k.metadata.options.parser_name'])
+
+        spec.outline(
+            cls.generate_displacements,
+            cls.collect_cp2k_inputs,
+            cls.run_cp2k_first,
+            while_(cls.should_run_displ)(cls.run_cp2k_displacements,),
+            cls.results,
+        )
+
+        # Ouptuts
+
+    def generate_displacements(self):
+        """Generate displacements using Phonopy"""
+
+        # CifData to PhonopyAtoms
+        ase = self.inputs.structure.get_ase()
+        uc = PhonopyAtoms(symbols=ase.get_chemical_symbols(),
+                          cell=ase.get_cell(),
+                          scaled_positions=ase.get_scaled_positions())
+
+        # Generate displacements
+        self.ctx.phonon = Phonopy(
+            unitcell=uc,
+            supercell_matrix=None,
+            primitive_matrix=None,
+            factor=CP2KToTHz,
+        )
+        self.ctx.phonon.generate_displacements(distance=0.01)
+        disp_ucs = self.ctx.phonon.supercells_with_displacements
+        #phonon.save() to investigate
+
+        # List of PhonopyAtoms to list of StructureData
+        self.ctx.sd_list = []
+        for i, pa in enumerate([uc] + disp_ucs):  # 6N+1 in total, first is the original uc
+            sd = StructureData(cell=uc.get_cell())
+            for i in range(len(uc)):
+                symbol = uc.get_chemical_symbols()[i]
+                position = uc.get_positions()[i]
+                sd.append_atom(symbols=symbol, position=position)
+            self.ctx.sd_list.append(sd)
+
+    def collect_cp2k_inputs(self):
+        """Collect Cp2k inputs from the last CP2K calculation."""
+        qb = QueryBuilder()
+        qb.append(Node, filters={f'id': self.inputs.structure.pk}, tag='cif_out')
+        qb.append(Node, filters={'attributes.process_label': 'Cp2kCalculation'}, with_descendants='cif_out', tag='calc')
```

Comment:
The idea of this is to query the latest CP2K calculation that produced the current Cif/Structure: but this comes from a previous WorkChain, with possible different connection. It is not a way to query a calculation that was done in this workchain.

Reply:
Oh, I see.  So you are querying for "the last" cp2k calculation that is somewhere up the provenance graph above the CIF?

I think this can have quite surprising results, and in principle this calculation does not need to have anything to do with the structure.
Also, what if the structure wasn't computed with cp2k?

I see that you chose to exclude the `cp2k.parameters` from the inputs.
I would suggest to add them back and to simply pass along the parameters of the relevant calculation in the input when calling this workchain.

Reply:
It does not make sense to compute the phonons, with displacements of 0.01 A, if the structure was not coming from a cp2k calculation, with exactly the same parameters. It would be cumbersome every time to search and load both parameters and possibly forgetting the files.
I think it is fine to keep it like this for the moment, it is very practical to use.

Reply:
Now I'm also trying to restart from the previous WFN just to save a little more time: it would be something more to add to the input (...I'm not even sure if a RemoteFolder can be passed as input).
It would be nice to pass the Cp2kCalculation as an input, but it is not a Data object.
Querying internally seems still the best solution to me.

Reply:
> I'm not even sure if a RemoteFolder can be passed as input

I think that should work

> It would be nice to pass the Cp2kCalculation as an input

You can always pass it by uuid

Reply:
The `uuid` of a CalcJob is maybe the best option.
Another problem it emerged (an I was expecting it!) is for those structures that have atoms with different ox state (like Cu1/Cu2 of `RSM0023` I'm reporting).
This info gets lost in the CifData and the calculation crashes because there are only Cu1/Cu2 kinds (which is better it fails so that we don't have some weird inconsistent output).
Therefore one want to use the CalcJob as input and take the output structure (if present, otherwise the input if the calculation is ENERGY) and keep the atom types from that StructureData

```
[<WorkChainNode: uuid: 661eb09c-314f-40b5-9f4b-1ae8bb72f663 (pk: 526519) (aiida_lsmo.workchains.cp2k_multistage_ddec.Cp2kMultistageDdecWorkChain)>,
 <Dict: uuid: 0bcbcb97-76c8-456d-8a3d-466dd2db0709 (pk: 534281)>,
 'RSM0023_DDEC',
 <CifData: uuid: 7294a4e5-d7dc-4a4c-9bb3-a8466632eb2d (pk: 538325)>]
 ```

Reply:
I implemented the Cp2kCalc.uuid as input in https://github.com/lsmo-epfl/aiida-lsmo/pull/79/commits/c1e12b76427cf7b86832a027d1b203fa2b2490fd, but I'm going to revert it to the previous choice of using a StructureData input, as I realized that having a completely disconnected WC (where the input is some newely created `Str` is very painful to manage: you should use dirty ways to check if you already submitted some calculation by including the material name in the label.